src/main.ts

```diff
@@ -156,33 +185,57 @@ ArgoCD Diff for commit [\`${shortCommitSha}\`](${commitLink})
   }
 }
 
+async function asyncForEach<T>(
+  array: T[],
+  callback: (item: T, i: number, arr: T[]) => Promise<void>
+): Promise<void> {
+  for (let index = 0; index < array.length; index++) {
+    await callback(array[index], index, array);
+  }
+}
+
 async function run(): Promise<void> {
   const argocd = await setupArgoCDCommand();
   const apps = await getApps();
   core.info(`Found apps: ${apps.map(a => a.metadata.name).join(', ')}`);
 
-  const diffPromises = apps.map(async app => {
+  const diffs: Diff[] = [];
+
+  await asyncForEach(apps, async app => {
     try {
       const command = `app diff ${app.metadata.name} --local=${app.spec.source.path}`;
-      if (app.spec.source.helm) {
-        const pwd = await execCommand(`pwd`);
-        core.info(`pwd: ${pwd.stdout}`);
-        const output = await execCommand(`cd ${app.spec.source.path} && ls -al`);
-        core.info(`output: ${output.stdout}`);
-      }
-      const res = await argocd(command);
-      core.info(`Running: argocd ${command}`);
-      core.info(`stdout: ${res.stdout}`);
-      core.info(`stdout: ${res.stderr}`);
-      if (res.stdout) {
-        return { app, diff: res.stdout };
+      try {
+        core.info(`Running: argocd ${command}`);
+        // ArgoCD app diff will exit 1 if there is a diff, so always catch,
+        // and then consider it a success if there's a diff in stdout
+        // https://github.com/argoproj/argo-cd/issues/3588
+        await argocd(command);
+      } catch (e) {
+        const res = e as ExecResult;
+        core.info(`stdout: ${res.stdout}`);
+        core.info(`stderr: ${res.stderr}`);
+        if (res.stdout) {
+          diffs.push({ app, diff: res.stdout });
+        } else {
+          diffs.push({
+            app,
+            diff: '',
+            error: `
+stderr: ${res.stderr}
+err: ${JSON.stringify(res.err)}
+          `
+          });
+        }
       }
     } catch (e) {
```

Comment:
What purpose does this outer `catch` serve, since you're not rethrowing any exceptions caught in the inner one?

Reply:
good point